File: typescript/packages/core/src/sandbox/protocol.ts
import type { EventOf } from "../fold/state";
import type { BranchId, SandboxId } from "../log";
import type { LookupCapability, LookupResult } from "../model/protocol";
import type { Result } from "../result";

// The sandbox adapter protocol (spec/api.json types Sandbox, SandboxInfo, SandboxSession,
// ExecOutput). Wire casing for data; option params are lowerCamel (api.json conventions).

/** What a snapshot records; the snapshot event's data. */
export type SnapshotData = EventOf<"snapshot">["data"];

/** An adapter's typed failure. Every code is a wire ErrorCode or an ApiErrorCode. */
export type Failure<Code extends string> = {
  readonly code: Code;
  readonly message: string;
};

export type SandboxInfo = {
  readonly provider: string;
  readonly egress: "enforced" | "unenforced";
  readonly capture_classes: readonly SnapshotData["capture_class"][];
  readonly browser: "none" | "headless";
  readonly desktop: "none" | "native" | "image";
  /** Per operation: create (Sandbox.lookup) and snapshot (Sandbox.lookupSnapshot). */
  readonly lookup: {
    readonly create: LookupCapability;
    readonly snapshot: LookupCapability;
  };
  /** unconfirmed: terminate always answers unknown, so every crash parks. */
  readonly termination: "confirmed" | "unconfirmed";
};

export type ExecOptions = {
  readonly cwd?: string;
  /** Exactly this env; nothing is inherited. */
  readonly env?: Readonly<Record<string, string>>;
  readonly timeoutMs?: number;
  readonly stdin?: Uint8Array;
  /** The call's effect key, durable in effect_begin before dispatch, so recovery can terminate it. */
  readonly processKey: string;
};

/** The complete output as byte streams, never truncated or buffered whole. */
export type ExecOutput = {
  /** Resolves once both streams end. */
  readonly exit_code: Promise<number>;
  readonly stdout: AsyncIterable<Uint8Array>;
  readonly stderr: AsyncIterable<Uint8Array>;
};

export type FileFailure = Failure<
  | "not_found"
  | "invalid_path"
  | "permission_denied"
  | "is_directory"
  | "too_large"
  | "unavailable"
>;

export type RestoreFailure = Failure<
  | "snapshot_expired"
  | "snapshot_missing"
  | "snapshot_restore_failed"
  | "snapshot_manifest_mismatch"
  | "unavailable"
>;

export type ReleaseFailure = Failure<"release_failed" | "unavailable">;

/**
 * A refused dispatch: the context's fence failed, so nothing reached the provider.
 * stale_epoch under owner authority, cleanup_claim_lost under cleanup authority.
 */
export type Stale = Failure<"stale_epoch" | "cleanup_claim_lost">;

/**
 * spec/api.json SandboxAuthority: the owner's lease at an epoch, or gc's cleanup claim on one
 * resources row (which outlives the owning branch).
 */
export type SandboxAuthority =
  | {
      readonly kind: "owner";
      readonly branch_id: BranchId;
      readonly epoch: number;
    }
  | {
      readonly kind: "cleanup";
      readonly resource_id: string;
      readonly claim: string;
    };

/**
 * spec/api.json SandboxContext: the authority an adapter re-checks at its real provider
 * dispatch point, after any queueing. A failed fence means call nothing.
 */
export type SandboxContext = {
  readonly authority: SandboxAuthority;
  readonly fence: () => Promise<Result<void, Stale>>;
};

export type SandboxSession = {
  readonly id: SandboxId;
  /** A timeout kills the whole process group. */
  readonly exec: (
    command: readonly string[],
    context: SandboxContext,
    options: ExecOptions,
  ) => Promise<
    Result<
      ExecOutput,
      Failure<"timeout" | "invalid_path" | "unavailable"> | Stale
    >
  >;
  /** Kill the process group started with this key and confirm it is gone. */
  readonly terminate: (
    processKey: string,
    context: SandboxContext,
  ) => Promise<
    Result<
      "terminated" | "already_exited" | "unknown",
      Failure<"unavailable"> | Stale
    >
  >;
  readonly upload: (
    path: string,
    data: Uint8Array,
    context: SandboxContext,
  ) => Promise<Result<void, FileFailure | Stale>>;
  readonly download: (
    path: string,
    context: SandboxContext,
  ) => Promise<Result<Uint8Array, FileFailure | Stale>>;
  /** Freeze or stop tracked processes, capture, thaw. Returns once durable and restorable. */
  readonly snapshot: (
    operationKey: string,
    context: SandboxContext,
  ) => Promise<
    Result<
      SnapshotData,
      Failure<"not_quiescent" | "unavailable" | "timeout"> | Stale
    >
  >;
  /** Release the sandbox: ok → released, an error → release_failed (retried by gc). */
  readonly close: (
    context: SandboxContext,
  ) => Promise<Result<void, ReleaseFailure | Stale>>;
} & Partial<Trees>;

/**
 * spec/api.json SandboxSession.exportTree and importTree: the optional capability to move
 * /workspace as one tar archive. Core reads an export with the tree reader (sandbox/trees.ts).
 */
export type Trees = {
  /** The archive on stdout, then exit code 0; any other code is a failed export. */
  readonly exportTree: (
    context: SandboxContext,
  ) => Promise<Result<ExecOutput, Failure<"unavailable"> | Stale>>;
  /** Extracts a host-built archive (modes already masked to 0o777) into /workspace, owner dropped. */
  readonly importTree: (
    tar: AsyncIterable<Uint8Array>,
    context: SandboxContext,
  ) => Promise<Result<void, Failure<"unavailable"> | Stale>>;
};

/** What a sandbox adapter returns. Every create and restore carries a ledgered operation key. */
export type Sandbox = {
  readonly info: SandboxInfo;
  /**
   * Resolves credentials and checks configuration on the host, at check() or the first run.
   * Opens no connection; a throw is a ConfigError, retried on the next check() or run.
   */
  readonly setup?: () => Promise<void>;
  readonly create: (
    operationKey: string,
    context: SandboxContext,
  ) => Promise<
    Result<SandboxSession, Failure<"unavailable" | "timeout"> | Stale>
  >;
  /**
   * Restores into a new isolated sandbox and verifies its tree against `manifestHash` (the
   * snapshot event's); a mismatch releases what it created and is snapshot_manifest_mismatch.
   */
  readonly restore: (
    snapshotId: string,
    manifestHash: string,
    operationKey: string,
    context: SandboxContext,
  ) => Promise<Result<SandboxSession, RestoreFailure | Stale>>;
  /** Present when info.lookup.create is not none. */
  readonly lookup?: (
    operationKey: string,
    context: SandboxContext,
  ) => Promise<Result<LookupResult<SandboxSession>, Stale>>;
  /** Present when info.lookup.snapshot is not none. */
  readonly lookupSnapshot?: (
    operationKey: string,
    context: SandboxContext,
  ) => Promise<Result<LookupResult<SnapshotData>, Stale>>;
  /** Reattach to a live row's sandbox by its ref, to terminate and close it after a restart. */
  readonly attach: (
    ref: string,
    context: SandboxContext,
  ) => Promise<
    Result<
      SandboxSession,
      Failure<"not_found" | "resource_unknown" | "unavailable"> | Stale
    >
  >;
  /** Release a snapshot by its ref. */
  readonly release: (
    ref: string,
    context: SandboxContext,
  ) => Promise<Result<"released" | "already_gone", ReleaseFailure | Stale>>;
};

File: python/src/threads/sandbox/protocol.py
"""The sandbox adapter protocol (spec/api.json `Sandbox`, `SandboxSession`, `SandboxInfo`,
`ExecOutput`, `ExecResult`).

Every create and restore carries an operation key that the resource ledger recorded first. Expected
failures are `SandboxError` values; an adapter raises only for bugs.
"""

import posixpath
from collections.abc import AsyncIterable, AsyncIterator, Awaitable, Mapping, Sequence
from dataclasses import dataclass
from types import MappingProxyType
from typing import Final, Literal, NewType, Protocol, assert_never, runtime_checkable

from threads.log import ArtifactRef, ParseError, SnapshotData
from threads.loop.model import LookupCapability, LookupResult, LookupUnknown
from threads.loop.tools import Termination
from threads.result import Err, Ok
from threads.store.context import CleanupAuthority, OwnerAuthority, SandboxAuthority

SandboxId = NewType("SandboxId", str)

type SandboxErrorCode = Literal[
    "unavailable",
    "timeout",
    "not_found",
    "invalid_path",
    "permission_denied",
    "is_directory",
    "too_large",
    "not_quiescent",
    "release_failed",
    "resource_unknown",
    "snapshot_expired",
    "snapshot_missing",
    "snapshot_restore_failed",
    "snapshot_manifest_mismatch",
    "stale_epoch",
    "cleanup_claim_lost",
]
"""The codes spec/api.json lists for the sandbox methods."""


@dataclass(frozen=True, slots=True)
class WorkspaceRefusal:
    """Why the thread's pinned workspace never reached a fresh sandbox (lane 16 E). Not an
    adapter failure: core places the tree, and the tool call that opened the session sees this
    instead of a bare "never sent"."""

    code: Literal["workspace_mismatch", "capability_missing"]
    message: str


type Refusal = Literal["stale_epoch", "cleanup_claim_lost"]
"""A failed fence: nothing reached the provider, and the caller lost its authority."""

NO_ENV: Final[Mapping[str, str]] = MappingProxyType({})


class SandboxContext(Protocol):
    """What the runtime hands an adapter for every provider operation: who
    dispatches it (an owner's lease, or gc's claim on the row), and `fence`, which the adapter
    awaits at its real dispatch point; a failure means call nothing and answer the `Refusal`
    its authority names."""

    @property
    def authority(self) -> SandboxAuthority: ...

    async def fence(self) -> Ok[None] | Err[ParseError]: ...


@dataclass(frozen=True, slots=True)
class SandboxError:
    code: SandboxErrorCode
    message: str


def refusal(authority: SandboxAuthority) -> Refusal:
    """stale_epoch under an owner's lease, cleanup_claim_lost under gc's claim."""
    match authority:
        case OwnerAuthority():
            return "stale_epoch"
        case CleanupAuthority():
            return "cleanup_claim_lost"
        case _:
            assert_never(authority)


def is_refusal(error: SandboxError) -> bool:
    return error.code in ("stale_epoch", "cleanup_claim_lost")


type Looked[T] = Ok[LookupResult[T]] | Err[SandboxError]
"""spec/api.json `Sandbox.lookup` and `lookupSnapshot` returns: the answer, or a refused fence
(stale_epoch or cleanup_claim_lost), when nothing was asked."""


def unanswered(error: SandboxError) -> Ok[LookupUnknown] | Err[SandboxError]:
    """A lookup whose provider call failed: a refused fence stays an error; any other failure
    is an answer that proves nothing."""
    return Err(error) if is_refusal(error) else Ok(LookupUnknown(f"{error.code}: {error.message}"))


async def refused(context: SandboxContext) -> Err[SandboxError] | None:
    """Checks the fence: None when it passed, else the typed refusal. An adapter calls it only at
    its real provider dispatch point."""
    passed = await context.fence()
    if isinstance(passed, Ok):
        return None
    return Err(SandboxError(refusal(context.authority), passed.error.message))


def invalid_path(path: str) -> Err[SandboxError] | None:
    """A sandbox path is absolute and already normal: no relative parts, no `..` escape."""
    if not path.startswith("/") or posixpath.normpath(path) != path:
        return Err(SandboxError("invalid_path", f"not an absolute normal path: {path}"))
    return None


@dataclass(frozen=True, slots=True)
class LookupSupport:
    """Per operation: create (`Sandbox.lookup`) and snapshot (`Sandbox.lookup_snapshot`)."""

    create: LookupCapability
    snapshot: LookupCapability


@dataclass(frozen=True, slots=True)
class SandboxInfo:
    provider: str
    egress: Literal["enforced", "unenforced"]
    capture_classes: tuple[Literal["filesystem", "filesystem_and_processes", "full_vm"], ...]
    browser: Literal["none", "headless"]
    desktop: Literal["none", "native", "image"]
    lookup: LookupSupport
    termination: Literal["confirmed", "unconfirmed"]
    """unconfirmed: every sandbox_local timeout or crash parks."""


@dataclass(frozen=True, slots=True)
class ExecOutput:
    """What an adapter's exec returns: the complete output as byte streams, never truncated
    or buffered whole. `exit_code` resolves once both streams end."""

    exit_code: Awaitable[int]
    stdout: AsyncIterator[bytes]
    stderr: AsyncIterator[bytes]


@dataclass(frozen=True, slots=True)
class ExecResult:
    """What tools and the log see, built by the sandbox layer from an `ExecOutput`."""

    exit_code: int
    stdout: str
    """Head and tail preview."""
    stderr: str
    """Head and tail preview."""
    truncated: bool
    full_output: ArtifactRef | None = None
    """Present exactly when truncated: the complete output, readable with read_tool_result."""


class SandboxSession(Protocol):
    @property
    def id(self) -> SandboxId: ...

    async def exec(  # noqa: PLR0913 - the options spec/api.json names
        self,
        command: Sequence[str],
        context: SandboxContext,
        *,
        process_key: str,
        cwd: str = "/workspace",
        env: Mapping[str, str] = NO_ENV,
        timeout_ms: int | None = None,
        stdin: bytes | None = None,
    ) -> Ok[ExecOutput] | Err[SandboxError]:
        """Runs with exactly `env`, nothing inherited. `process_key` (the call's effect key) is
        the durable identity of the process group, so recovery can terminate it. A timeout
        kills the whole group."""
        ...

    async def terminate(
        self, process_key: str, context: SandboxContext
    ) -> Ok[Termination] | Err[SandboxError]: ...

    async def upload(
        self, path: str, data: bytes, context: SandboxContext
    ) -> Ok[None] | Err[SandboxError]: ...

    async def download(
        self, path: str, context: SandboxContext
    ) -> Ok[bytes] | Err[SandboxError]: ...

    async def snapshot(
        self, operation_key: str, context: SandboxContext
    ) -> Ok[SnapshotData] | Err[SandboxError]:
        """Freezes or stops tracked processes, captures, thaws. Returns once durable."""
        ...

    async def close(self, context: SandboxContext) -> Ok[None] | Err[SandboxError]:
        """Releases the sandbox."""
        ...


class Sandbox(Protocol):
    """spec/api.json `Sandbox`. A sandbox whose `info.lookup.create` is not none also
    implements `LooksUpSandbox`, and one whose `info.lookup.snapshot` is not none
    `LooksUpSnapshot`; check() and the first run refuse one that doesn't."""

    @property
    def info(self) -> SandboxInfo: ...

    async def create(
        self, operation_key: str, context: SandboxContext
    ) -> Ok[SandboxSession] | Err[SandboxError]: ...

    async def restore(
        self, snapshot_id: str, manifest_hash: str, operation_key: str, context: SandboxContext
    ) -> Ok[SandboxSession] | Err[SandboxError]:
        """Restores into a new isolated sandbox and verifies it: a restored tree whose
        canonical manifest hash isn't `manifest_hash` is snapshot_manifest_mismatch, and the
        adapter releases the sandbox it created first."""
        ...

    async def attach(
        self, ref: str, context: SandboxContext
    ) -> Ok[SandboxSession] | Err[SandboxError]:
        """Reattaches to a live row's sandbox by its recorded ref."""
        ...

    async def release(
        self, ref: str, context: SandboxContext
    ) -> Ok[Literal["released", "already_gone"]] | Err[SandboxError]:
        """Releases a snapshot by its durable ref."""
        ...


@runtime_checkable
class Trees(Protocol):
    """spec/api.json `SandboxSession.export_tree` and `import_tree`, an optional capability of a
    `SandboxSession`: /workspace moved as one tar archive. Core reads every export with the
    strict tree reader (sandbox/trees.py)."""

    async def export_tree(self, context: SandboxContext) -> Ok[ExecOutput] | Err[SandboxError]:
        """The archive on stdout, then exit code 0; any other code is a failed export."""
        ...

    async def import_tree(
        self, tar: AsyncIterable[bytes], context: SandboxContext
    ) -> Ok[None] | Err[SandboxError]:
        """Extracts a host-built archive (modes already masked to 0o777) into /workspace, owner
        dropped."""
        ...


@runtime_checkable
class LooksUpSandbox(Protocol):
    """spec/api.json `Sandbox.lookup`, an optional capability of a `Sandbox`."""

    async def lookup(self, operation_key: str, context: SandboxContext) -> Looked[SandboxSession]:
        """Finds a sandbox whose create or restore answer was lost."""
        ...


@runtime_checkable
class LooksUpSnapshot(Protocol):
    """spec/api.json `Sandbox.lookupSnapshot`, an optional capability of a `Sandbox`."""

    async def lookup_snapshot(
        self, operation_key: str, context: SandboxContext
    ) -> Looked[SnapshotData]:
        """Finds a snapshot whose capture answer was lost."""
        ...
